qutip/solver/ode/krylov.py

```diff
@@ -0,0 +1,236 @@
+                self._max_step = self._compute_max_step(krylov_tridiag,
+                                                        krylov_basis)
+
+    def _lanczos_algorithm(self, psi):
+        """
+        Computes a basis of the Krylov subspace for Hamiltonian 'H', a system
+        state 'psi' and Krylov dimension 'krylov_dim'. The space is spanned
+        by {psi, H psi, H^2 psi, ..., H^(krylov_dim) psi}.
+
+        Parameters
+        ------------
+        psi: np.ndarray
+            State used to calculate Krylov subspace.
+        """
+        krylov_dim = self.options['krylov_dim']
+        H = (1j * self.system(0)).data
+
+        v = []
+        T_diag = np.zeros(krylov_dim + 1, dtype=complex)
+        T_subdiag = np.zeros(krylov_dim + 1, dtype=complex)
+
+        w_prime = _data.matmul(H, psi)
+        T_diag[0] = _data.inner(w_prime, psi)
+        v.append(psi)
+        w = _data.add(w_prime, v[-1], -T_diag[0])
+        T_subdiag[0] = _data.norm.l2(w)
+        j = 0
+
+        while j < krylov_dim and T_subdiag[j] > self.options['sub_system_tol']:
+            j += 1
+            v.append(_data.mul(w, 1 / T_subdiag[j-1]))
+            w_prime = _data.matmul(H, v[-1])
+            T_diag[j] = _data.inner(w_prime, v[-1])
+            w = _data.add(w_prime, v[-1], -T_diag[j])
+            w = _data.add(w, v[-2], -T_subdiag[j-1])
+            T_subdiag[j] = _data.norm.l2(w)
+
+        krylov_tridiag = _data.diag["dense"](
+            [T_subdiag[:j], T_diag[:j+1], T_subdiag[:j]],
+            [-1, 0, 1]
+        )
+        krylov_basis = _data.Dense(np.hstack([psi.to_array() for psi in v]))
+
+        return krylov_tridiag, krylov_basis
+
+    def _compute_krylov_set(self, krylov_tridiag, krylov_basis):
+        """
+        Compute the eigen energies, basis transformation operator (U) and e0.
+        """
+        eigenvalues, eigenvectors = _data.eigs(krylov_tridiag, True)
+        N = eigenvalues.shape[0]
+        U = _data.matmul(krylov_basis, eigenvectors)
+        e0 = eigenvectors.adjoint() @ _data.one_element_dense((N, 1), (0, 0), 1.0)
+        return eigenvalues, U, e0
+
+    def _compute_psi(self, dt, eigenvalues, U, e0):
+        """
+        compute the state at time ``t``.
+        """
+        phases = _data.Dense(np.exp(-1j * dt * eigenvalues))
+        aux = _data.multiply(phases, e0)
+        return _data.matmul(U, aux)
+
+    def _compute_max_step(self, krylov_tridiag, krylov_basis, krylov_state=None):
+        """
+        Compute the maximum step length to stay under the desired tolerance.
+        """
+        if not krylov_state:
+            krylov_state = self._compute_krylov_set(krylov_tridiag, krylov_basis)
+
+        small_tridiag = _data.Dense(krylov_tridiag.as_ndarray()[:-1, :-1])
+        small_basis = _data.Dense(krylov_basis.as_ndarray()[:, :-1])
+        reduced_state = self._compute_krylov_set(small_tridiag, small_basis)
+
+        def krylov_error(t):
+            return np.log(_data.norm.l2(
```

Comment:
```suggestion
            # we divide by atol and take the log so that the error returned is 0
            # at atol, which is convenient for calling root_scalar with.
            return np.log(_data.norm.l2(
```